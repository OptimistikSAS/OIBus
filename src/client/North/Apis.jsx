import OIConnect from '../../north/oiconnect/OIConnect.schema.jsx'
import AmazonS3 from '../../north/amazon/AmazonS3.schema.jsx'
import Console from '../../north/console/Console.schema.jsx'
import InfluxDB from '../../north/influxdb/InfluxDB.schema.jsx'
import OIAnalyticsFile from '../../north/oianalyticsfile/OIAnalyticsFile.schema.jsx'
import TimescaleDB from '../../north/timescaledb/TimescaleDB.schema.jsx'

const ApiSchemas = { OIConnect, AmazonS3, Console, InfluxDB, OIAnalyticsFile, TimescaleDB }

export default ApiSchemas
