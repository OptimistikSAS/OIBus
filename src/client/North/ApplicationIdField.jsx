import React from 'react'
import PropTypes from 'prop-types'
import { Button } from 'reactstrap'
import { FaPencilAlt, FaCheck } from 'react-icons/fa'
import { OIbText } from '../components/OIbForm'

const ApplicationIdField = ({ applicationId, applications, applicationIndex, applicationIdChanged }) => {
  const [editing, setEditing] = React.useState(false)
  const [editingApplicationId, setEditingApplicationId] = React.useState()
  const [otherApplications, setOtherApplications] = React.useState()
  const [editingError, setEditingError] = React.useState()

  const handleEditName = () => {
    setEditing(true)
    setEditingApplicationId(applicationId)
    const applicationsList = applications.filter((e, i) => i !== applicationIndex).map((application) => application.applicationId)
    setOtherApplications(applicationsList)
  }

  const isValidName = (val) => {
    let error = null
    if (otherApplications.includes(val)) {
      error = 'Application id already exists'
    }
    if (!error) {
      error = (((typeof val === 'string' || val instanceof String) && val !== '') ? null : 'value must not be empty')
    }
    setEditingError(error)
    return error
  }

  const onChange = (name, value) => {
    setEditingApplicationId(value)
  }

  const handleDoneEditName = () => {
    if (!editingError) {
      applicationIdChanged(applicationIndex, editingApplicationId)
      setEditing(false)
    }
  }

  const style = { display: 'inline-block' }
  return (
    editing ? (
      <div style={style}>
        <OIbText
          onChange={onChange}
          value={editingApplicationId}
          valid={isValidName}
          name="applicationId"
          inline
        />
        <Button close>
          <FaCheck
            className="oi-icon oi-icon-inline"
            onClick={handleDoneEditName}
          />
        </Button>
      </div>
    ) : (
      <div style={style}>
        {applicationId}
        <Button close>
          <FaPencilAlt
            className="oi-icon oi-icon-inline"
            onClick={() => {
              handleEditName()
            }}
            style={{ height: 12 }}
          />
        </Button>
      </div>
    )
  )
}

ApplicationIdField.propTypes = {
  applicationId: PropTypes.string.isRequired,
  applications: PropTypes.arrayOf(Object).isRequired,
  applicationIndex: PropTypes.number.isRequired,
  applicationIdChanged: PropTypes.func.isRequired,
}
export default ApplicationIdField
