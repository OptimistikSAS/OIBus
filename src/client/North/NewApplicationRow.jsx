import React from 'react'
import PropTypes from 'prop-types'
import { Button, Form, Col, Row } from 'reactstrap'
import { OIbText, OIbSelect } from '../components/OIbForm/index'

const NewApplicationRow = ({ apiList, addApplication }) => {
  const [applicationId, setApplicationId] = React.useState('')
  const [api, setApi] = React.useState(apiList[0])
  /**
   * Updates the application's state
   * @param {*} event The change event
   * @returns {void}
   */
  const handleAddApplication = () => {
    //  update the new application's state
    if (applicationId === '') return
    addApplication({ applicationId, api })
  }

  const handleChange = (name, value) => {
    switch (name) {
      case 'applicationId':
        setApplicationId(value)
        break
      case 'api':
      default:
        setApi(value)
        break
    }
  }
  return (
    <Form>
      <Row>
        <Col md="5">
          <OIbText
            label="New Application ID"
            value={applicationId}
            name="applicationId"
            onChange={handleChange}
            defaultValue=""
          />
        </Col>
        <Col md="3">
          <OIbSelect label="API" value={api} name="api" options={apiList} defaultValue={apiList[0]} onChange={handleChange} />
        </Col>
        <Col md="3">
          <Button size="sm" className="oi-add-button" color="primary" onClick={() => handleAddApplication()}>
            Add
          </Button>
        </Col>
      </Row>
    </Form>
  )
}

NewApplicationRow.propTypes = {
  apiList: PropTypes.arrayOf(PropTypes.string).isRequired,
  addApplication: PropTypes.func.isRequired,
}
export default NewApplicationRow
