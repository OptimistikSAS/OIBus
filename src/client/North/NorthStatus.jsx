import React from 'react'
import { Container, Button } from 'reactstrap'
import { useParams, useNavigate } from 'react-router-dom'
import { FaArrowLeft } from 'react-icons/fa'
import { ConfigContext } from '../context/ConfigContext.jsx'
import Table from '../components/table/Table.jsx'
import utils from '../helpers/utils'

/**
 * Generate row entry for the status table.
 * @param {string} key - The key
 * @param {string} value - The value
 * @return {[{name: *, value: *}, {name: string, value: *}]} - The table row
 */
const generateRowEntry = (key, value) => [
  {
    name: key,
    value: key,
  },
  {
    name: 'value',
    value: utils.formatValue(value),
  },
]

const NorthStatus = () => {
  const [connectorData, setConnectorData] = React.useState([])
  const { newConfig } = React.useContext(ConfigContext)
  const navigate = useNavigate()
  const { id } = useParams() // the application id passed in the url
  const [application, setApplication] = React.useState(null)

  React.useEffect(() => {
    const currentApplication = newConfig.north?.applications?.find(
      (element) => element.id === id,
    )
    setApplication(currentApplication)

    let source
    if (currentApplication && currentApplication.enabled) {
      source = new EventSource(`/north/${id}/sse`)
      source.onerror = (error) => {
        console.error(error)
      }
      source.onmessage = (event) => {
        if (event && event.data) {
          const myData = JSON.parse(event.data)
          const tableRows = []
          Object.keys(myData).forEach((key) => {
            tableRows.push(generateRowEntry(key, myData[key]))
          })
          setConnectorData(tableRows)
        }
      }
    }
    return () => {
      source?.close()
    }
  }, [newConfig])

  return application ? (
    <>
      <div
        id="oi-sub-nav"
        className="d-flex align-items-center w-100 oi-sub-nav mb-2"
      >
        <h6 className="text-muted d-flex align-items-center ps-3 pt-1">
          <Button
            id="oi-navigate"
            outline
            onClick={() => {
              navigate(-1)
            }}
            className="util-button"
          >
            <FaArrowLeft className="oi-back-icon me-2" />
          </Button>
          {`| ${application.name}`}
        </h6>
      </div>
      <Container>
        {connectorData.length > 0 ? (
          <Table headers={[]} rows={connectorData} />
        ) : (
          <div className="oi-status-error">
            {' '}
            No Information displayed because the connector is disabled
            {' '}
          </div>
        )}
      </Container>
    </>
  ) : null
}
export default NorthStatus
