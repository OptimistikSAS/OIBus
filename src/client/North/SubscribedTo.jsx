import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'
import Table from '../components/table/Table.jsx'
import { OIbSelect, OIbTitle } from '../components/OIbForm'
import { ConfigContext } from '../context/configContext.jsx'

const SubscribedTo = ({ subscribedTo }) => {
  const { dispatchNewConfig, newConfig } = React.useContext(ConfigContext)
  const dataSourceIds = newConfig && newConfig.south.dataSources.map((dataSource) => dataSource.dataSourceId)
  const handleDelete = (rowIndex) => {
    dispatchNewConfig({ type: 'deleteRow', name: `engine.filter.${rowIndex}` })
  }
  const handleAdd = () => {
    dispatchNewConfig({ type: 'addRow', name: 'engine.filter', value: '' })
  }
  const onChange = (name, value, validity) => {
    dispatchNewConfig({ type: 'update', name, value, validity })
  }
  return (
    subscribedTo && (
      <>
        <OIbTitle title="Subscribed To">
          <p>allow to select South equipment (default is to receive from all enabled equipment of the current OIBus </p>
        </OIbTitle>
        <Row>
          <Col md={4}>
            <Table
              headers={['Data source']}
              rows={subscribedTo.map((dataSource, i) => [
                {
                  name: `subscribedTo.${i}`,
                  value: (
                    <OIbSelect
                      name={`subscribedTo.${i}`}
                      option={subscribedTo[i]}
                      options={dataSourceIds}
                      onChange={onChange}
                    />
                  ),
                },
              ])}
              handleDelete={handleDelete}
              handleAdd={handleAdd}
            />
          </Col>
        </Row>
      </>
    )
  )
}

SubscribedTo.propTypes = { subscribedTo: PropTypes.arrayOf(String).isRequired }

export default SubscribedTo
