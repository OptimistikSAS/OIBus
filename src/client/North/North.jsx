import React from 'react'
import { useHistory, Link } from 'react-router-dom'
import { Col, Spinner, Breadcrumb, BreadcrumbItem } from 'reactstrap'
import Table from '../components/table/Table.jsx'
import NewApplicationRow from './NewApplicationRow.jsx'
import { AlertContext } from '../context/AlertContext.jsx'
import { ConfigContext } from '../context/configContext.jsx'
import EditableIdField from '../components/EditableIdField.jsx'

const North = () => {
  const { setAlert } = React.useContext(AlertContext)
  const { newConfig, dispatchNewConfig, apiList } = React.useContext(ConfigContext)
  const applications = newConfig?.north?.applications
  const history = useHistory()

  /**
   * Gets the config json of a north application
   * @param {string} applicationId ID of an application
   * @returns {object} The selected application's config
   */
  const getApplicationIndex = (applicationId) => applications.findIndex((application) => application.applicationId === applicationId)

  /**
   * Handles the edit of application and redirects the
   * user to the selected north applications's configuration page
   * @param {integer} index The id to edit
   * @return {void}
   */
  const handleEdit = (index) => {
    const application = applications[index]
    const link = `/north/${application.applicationId}`
    history.push({ pathname: link })
  }

  /**
   * Handles the change of one application id (name)
   * @param {integer} applicationIndex The index that will change
   * @param {string} newApplicationId The new application id
   * @return {void}
   */
  const handleApplicationIdChanged = (applicationIndex, newApplicationId) => {
    dispatchNewConfig({
      type: 'update',
      name: `north.applications.${applicationIndex}.applicationId`,
      value: newApplicationId,
    })
  }

  /**
   * Adds a new application row to the table
   * @param {Object} param0 An application object containing
   * applicationId, enabled and api fields
   * @returns {void}
   */
  const addApplication = ({ applicationId, api }) => {
    const applicationIndex = getApplicationIndex(applicationId)
    if (applicationIndex === -1) {
      // Adds new application
      dispatchNewConfig({ type: 'addRow', name: 'north.applications', value: { applicationId, api, enabled: false } })
    } else {
      const error = new Error('application already exists')
      setAlert({ text: error.message, type: 'danger' })
      throw error
    }
  }

  /**
   * Deletes the chosen application
   * @param {integer} index The id to delete
   * @returns {void}
   */
  const handleDelete = (index) => {
    dispatchNewConfig({ type: 'deleteRow', name: `north.applications.${index}` })
  }

  const tableHeaders = ['Application ID', 'Status', 'API']
  const tableRows = applications?.map(({ applicationId, enabled, api }, index) => [
    {
      name: applicationId,
      value: (
        <EditableIdField
          id={applicationId}
          fromList={applications}
          index={index}
          name="applicationId"
          idChanged={handleApplicationIdChanged}
        />
      ),
    },
    {
      name: 'enabled',
      value: <div className={enabled ? 'text-success' : 'text-danger'}>{enabled ? 'Enabled' : 'Disabled'}</div>,
    },
    { name: 'api', value: api },
  ])

  return tableRows && Array.isArray(apiList) ? (
    <Col md="6">
      <Breadcrumb tag="h5">
        <BreadcrumbItem tag={Link} to="/" className="oi-breadcrumb">
          Home
        </BreadcrumbItem>
        <BreadcrumbItem active tag="span">
          North
        </BreadcrumbItem>
      </Breadcrumb>
      <Table headers={tableHeaders} rows={tableRows} handleEdit={handleEdit} handleDelete={handleDelete} />
      <NewApplicationRow apiList={apiList} addApplication={addApplication} />
    </Col>
  ) : (
    <div className="spinner-container">
      <Spinner color="primary" type="grow" />
      ...loading configuration from OIBus server...
    </div>
  )
}

export default North
