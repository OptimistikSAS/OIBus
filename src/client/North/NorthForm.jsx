import React from 'react'
import PropTypes from 'prop-types'
import { Form, Row, Col } from 'reactstrap'
// import { AlertContext } from '../context/AlertContext.jsx'
import { OIbTitle, OIbCheckBox, OIbInteger } from '../components/OIbForm'
import SubscribedTo from './SubscribedTo.jsx'

import Link from '../../north/link/Link.Form.jsx'

const ApiForms = { Link }

const NorthForm = ({ application, onChange }) => {
  const { api, applicationId } = application
  const ApiForm = ApiForms[api]
  return (
    <Form>
      <OIbTitle title={`${applicationId} parameters (api: ${api})`}>
        <>
          <ul>
            <li>...</li>
          </ul>
        </>
      </OIbTitle>
      <Row>
        <Col md={2}>
          <OIbCheckBox name="enabled" label="Enabled" value={application.enabled} help={<div>Enable this application</div>} onChange={onChange} />
        </Col>
      </Row>
      <ApiForm onChange={onChange} application={application} />
      <OIbTitle title="Caching" />
      <Row>
        <Col md="4">
          <OIbInteger
            onChange={onChange}
            value={application.caching.sendInterval}
            name="caching.sendInterval"
            help={<div>Value in milliseconds for data sending interval</div>}
          />
        </Col>
        <Col md="4">
          <OIbInteger
            onChange={onChange}
            value={application.caching.retryInterval}
            name="caching.retryInterval"
            help={<div>Value in milliseconds for retry sending data in case of failure</div>}
          />
        </Col>
      </Row>
      <Row>
        <Col md="4">
          <OIbInteger
            onChange={onChange}
            value={application.caching.groupCount}
            name="caching.groupCount"
            help={<div>The minimum buffer that will ensure date is not sent until value is reached</div>}
          />
        </Col>
        <Col md="4">
          <OIbInteger onChange={onChange} value={application.caching.maxSendCount} name="caching.maxSendCount" help={<div />} />
        </Col>
      </Row>
      <SubscribedTo onChange={onChange} subscribedTo={application.subscribedTo || []} />
    </Form>
  )
}

NorthForm.propTypes = { application: PropTypes.object.isRequired, onChange: PropTypes.func.isRequired }

export default NorthForm
