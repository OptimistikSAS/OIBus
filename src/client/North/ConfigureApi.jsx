import React from 'react'
import { useParams } from 'react-router-dom'
import { Spinner } from 'reactstrap'
import { ConfigContext } from '../context/configContext.jsx'
import NorthForm from './Form/NorthForm.jsx'

const ConfigureApi = () => {
  const { newConfig, dispatchNewConfig } = React.useContext(ConfigContext)
  const applications = newConfig?.north?.applications // array of all defined applications
  const { applicationId } = useParams()
  const applicationIndex = applications?.findIndex((application) => application.applicationId === applicationId)

  const onChange = (name, value, validity) => {
    dispatchNewConfig({ type: 'update', name, value, validity })
  }
  return applicationIndex >= 0 ? (
    <NorthForm
      application={applications[applicationIndex]}
      applicationIndex={applicationIndex}
      onChange={onChange}
    />
  ) : (
    <div className="spinner-container">
      <Spinner color="primary" type="grow" />
      ...loading configuration from OIBus server...
    </div>
  )
}

export default ConfigureApi
