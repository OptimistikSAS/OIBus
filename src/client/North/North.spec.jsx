import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import newConfig from '../../../tests/testConfig'
import North from './North.jsx'

React.useContext = jest.fn().mockReturnValue({ newConfig, apiList: newConfig.apiList })

jest.mock('react-router-dom', () => (
  { useHistory: jest.fn().mockReturnValue([]) }
))

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('North', () => {
  test('check North', () => {
    act(() => {
      ReactDOM.render(
        <North />, container,
      )
    })
    expect(container).toMatchSnapshot()
  })
})
