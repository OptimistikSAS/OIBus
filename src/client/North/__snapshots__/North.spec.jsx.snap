// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`North check North 1`] = `
<div>
  <div
    class="col-md-6"
  >
    <h5
      aria-label="breadcrumb"
      class=""
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="oi-breadcrumb breadcrumb-item"
          to="/"
        >
          Home
        </li>
        <span
          aria-current="page"
          class="active breadcrumb-item"
        >
          North
        </span>
      </ol>
    </h5>
    <table
      class="table table-sm table-striped table-hover"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            Application ID
          </th>
          <th
            scope="col"
          >
            Status
          </th>
          <th
            scope="col"
          >
            API
          </th>
          <th
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <div
              style="display: inline-block;"
            >
              c
              <button
                aria-label="Close"
                class="close"
              >
                <svg
                  class="oi-icon oi-icon-inline"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="height: 16px;"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                  />
                </svg>
              </button>
            </div>
          </td>
          <td>
            <div
              class="text-success"
            >
              Enabled
            </div>
          </td>
          <td>
            Console
          </td>
          <td>
            <svg
              class="oi-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              />
            </svg>
            <svg
              class="oi-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
              />
            </svg>
          </td>
        </tr>
        <tr>
          <td>
            <div
              style="display: inline-block;"
            >
              monoiconnect
              <button
                aria-label="Close"
                class="close"
              >
                <svg
                  class="oi-icon oi-icon-inline"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  style="height: 16px;"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                  />
                </svg>
              </button>
            </div>
          </td>
          <td>
            <div
              class="text-danger"
            >
              Disabled
            </div>
          </td>
          <td>
            OIConnect
          </td>
          <td>
            <svg
              class="oi-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
              />
            </svg>
            <svg
              class="oi-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
              />
            </svg>
          </td>
        </tr>
      </tbody>
    </table>
    <form
      class=""
    >
      <div
        class="row"
      >
        <div
          class="col-md-5"
        >
          <div
            class="form-group"
          >
            <label
              class=""
              for="applicationId"
            >
              New Application ID
            </label>
            <input
              class="oi-form-input form-control"
              id="applicationId"
              name="applicationId"
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-3"
        >
          <div
            class="form-group"
          >
            <label
              class=""
              for="api"
            >
              API
            </label>
            <select
              class="oi-form-input form-control"
              id="api"
              name="api"
            >
              <option
                value="Console"
              >
                Console
              </option>
              <option
                value="OIConnect"
              >
                OIConnect
              </option>
            </select>
            <div
              class="invalid-feedback"
            />
          </div>
        </div>
        <div
          class="col-md-3"
        >
          <button
            class="oi-add-button btn btn-primary btn-sm"
            type="button"
          >
            Add
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
